google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1alpha2/JsonToProtoMessageTest.java

```diff
@@ -0,0 +1,638 @@
+        break;
+      case INT64:
+        assertTrue((long) value == json.getLong(fieldName));
+        break;
+      case INT32:
+        assertTrue((int) value == json.getInt(fieldName));
+        break;
+      case STRING:
+        assertTrue(((String) value).equals(json.getString(fieldName)));
+        break;
+      case DOUBLE:
+        assertTrue((double) value == json.getDouble(fieldName));
+        break;
+      case MESSAGE:
+        AreMatchingFieldsFilledIn((DynamicMessage) value, json.getJSONObject(fieldName));
+        break;
+    }
+  }
+
+  private void isProtoArrayJsonArrayEqual(
+      FieldDescriptor key,
+      java.lang.Object value,
+      JSONObject json,
+      HashMap<String, String> jsonLowercaseNameToActualName) {
+    String fieldName = jsonLowercaseNameToActualName.get(key.getName().toLowerCase());
+    JSONArray jsonArray = json.getJSONArray(fieldName);
+    switch (key.getType()) {
+      case BOOL:
+        List<Boolean> boolArr = (List<Boolean>) value;
+        for (int i = 0; i < jsonArray.length(); i++) {
+          assertTrue((boolArr.get(i) == jsonArray.getBoolean(i)));
+        }
+        break;
+      case BYTES:
+        List<byte[]> byteArr = (List<byte[]>) value;
+        for (int i = 0; i < jsonArray.length(); i++) {
+          assertTrue(Arrays.equals(byteArr.get(i), jsonArray.getString(i).getBytes()));
+        }
+        break;
+      case INT64:
+        List<Long> longArr = (List<Long>) value;
+        for (int i = 0; i < jsonArray.length(); i++) {
+          assertTrue((longArr.get(i) == jsonArray.getLong(i)));
+        }
+        break;
+      case INT32:
+        List<Integer> intArr = (List<Integer>) value;
+        for (int i = 0; i < jsonArray.length(); i++) {
+          assertTrue((intArr.get(i) == jsonArray.getInt(i)));
+        }
+        break;
+      case STRING:
+        List<String> stringArr = (List<String>) value;
+        for (int i = 0; i < jsonArray.length(); i++) {
+          assertTrue(stringArr.get(i).equals(jsonArray.getString(i)));
+        }
+        break;
+      case DOUBLE:
+        List<Double> doubleArr = (List<Double>) value;
+        for (int i = 0; i < jsonArray.length(); i++) {
+          assertTrue((doubleArr.get(i) == jsonArray.getDouble(i)));
+        }
+        break;
+      case MESSAGE:
+        List<DynamicMessage> messageArr = (List<DynamicMessage>) value;
+        for (int i = 0; i < jsonArray.length(); i++) {
+          AreMatchingFieldsFilledIn(messageArr.get(i), jsonArray.getJSONObject(i));
+        }
+        break;
+    }
+  }
+
+  @Test
+  public void testDifferentNameCasing() throws Exception {
+    JSONObject json = new JSONObject();
+    json.put("bYtE", (byte) 1);
+    json.put("SHORT", (short) 1);
+    json.put("inT", 1);
+    json.put("lONg", 1L);
+    DynamicMessage protoMsg =
+        JsonToProtoMessage.convertJsonToProtoMessage(TestInt64.getDescriptor(), json, false);
+    AreMatchingFieldsFilledIn(protoMsg, json);
```

Comment:
Seems we are using a function "AreMatchingFieldsFilledIn" to verify whether proto equals to json. What if there's a bug inside AreMatchingFieldsFilledIn?.

Can we assert protoMsg to a pre-defined proto?

Reply:
The proto message produced does use a pre-defined proto, so it must be of the type of the pre-defined proto. For example, for whatever protoMsg produced, if we call the function getDescriptorForType(), it will be equal to the protobuf descriptor that was passed in to create it. The AreMatchingFieldsFilledIn() function tests for field existence and if the field value is the same as the json value.

Reply:
Changed testcases to compare generated proto message with self defined proto message.